chain/messagepool/messagepool.go

```diff
@@ -449,8 +470,8 @@ func New(ctx context.Context, api Provider, ds dtypes.MetadataDS, us stmgr.Upgra
 }
 
 func (mp *MessagePool) ForEachPendingMessage(f func(cid.Cid) error) error {
-	mp.lk.Lock()
-	defer mp.lk.Unlock()
+	mp.transactionLk.Lock()
```

Comment:
I think it's sufficient to Lock mp.stateLk here, since we just need `pending` to not change under our feet?

Reply:
Transaction lock needs to be used to safely make changes to state. So changes to the state must be wrapped in a transaction lock. Hope that makes sense! Also the protector is who calls this function so I am pretty sure it wants write access not read access, so i used a write lock here